### DISCUSSION TITLE
補充說明 虛擬環境的建立& requirement.txt 建立原因

### DISCUSSION THREAD

**Discussion by Bill671** [1101-GS3073-C/Python-Community-GS3073-C#281] [General]
## 前提概要
上課時，老師有分別教大家用ananconda 圖像介面跟用原生cpython創建虛擬環境，
虛擬環境好處因為已經有其他大神介紹過了，這邊就不多說，只重點補充幾點
* 套件版本相容性的最佳解決方案
> 每個套件相容其他套件的版本可能不同，比如從**PDF提取文字的camelot**套件相容的pandas版本比較舊，
( 更準確來說，是新一輪pandas使用較新的xlrd套件版本 )
但是今天你想要使用pandas新一版本的功能好比 read_csv或read_xlsx，這樣就會報錯，想要讀取文字又想要使用新一版本pandas功能，虛擬環境的建置就成了很好的解決方案
>
* 想要獨立開發專案
> 今天如果想要獨立開發並包裝成exe或者其他程式檔案，我會希望打包的套件程式能越少越好，這樣exe檔案才能夠越小，這時候虛擬環境就可以協助你讓你有個獨立開發空間去開發

## 安裝補充說明
其實anaconda 本身除了用圖形UI介面建置虛擬環境外，
他還提供更快速的建制方式，打開Anaconda powershell Prompt，他是anaconda版的cmd，
一定要使用powershell版本!!
這是因為這樣才是系統管理員執行，不然指令無法輸出 ! 開啟 Anaconda Powershell Prompt後輸入
`conda create --name your_env_name python=3.7` 這樣的格式，
意思是建立一個名稱為your_env_name的python 3.7版本的虛擬環境，
你也可以輸入`conda create --n your_env_name python=3.7`，兩者意思相同
按下enter 後就建置完畢，接下來如果我想要啟用這個名稱要怎麼做呢?
在anaconda powershell prompt 輸入 `conda activate your_env_name`，這樣就進入到你建置的環境了
>
接下來你可能會問，**這樣建置跟老師用python原生方式建置或者用anaconda 圖像建置的差異在哪?**
>
首先，你可以在powershell prompt 輸入 `conda env list` 查看你的虛擬環境有幾個，這是原生方式無法做到的
(除非你虛擬環境在anaconda env底下)
anaconda 圖像UI方式建置雖然也可以做到，但建置時間比起我們輸入指令建置來的更慢，
用 %timeit來說就是建置時間完全不在一個量級
>
其次，activate 方式就直接輸入 `conda activate your_env`，而不用去找他的activate.bat檔案
這對使用者來說會便利很多
>
最後，想補充一下查看函式指令，除了老師所說的 pip freeze，另一個查看函式包的指令是 pip list，也是我比較常用的方式，兩者的差別在於 pip freeze不會顯示原先python依存的套件，也就是你想刪但無法刪(我不太確定是不是無法刪，但一定是比較難刪除)，而pip list除了列出你安裝的套件外，他也會列出原先python依存的套件
>
另外，可能大家並不知道為什麼老師要示範 pip freeze >requirement.txt，這是因為可以透過 requirement.txt 來安裝套件，比如我覺得我想搬移我的環境，那我可以先pip freeze > requirements.txt 匯出我的環境套件，再到新建立的環境中輸入
` pip install -r requirements.txt ` 來安裝我的套件

( 印象中沒記錯指令 -r 代表 read 讀取，有說錯歡迎補充說明 )

**Comment by lx3211** [comment]
還以為requirement.txt是用來部署才會用到，原來到新環境也可以用指令來一次下載，真方便

**Comment by xCk27x** [comment]
老師上課時聽得沒有很懂，看完這篇後，對requirement.txt的了解多了很多

**Comment by limitx0** [comment]
感謝超詳細的說明，讓我學到更多
